src/main/java/com/github/manolo8/darkbot/extensions/plugins/PluginUpdater.java

```diff
@@ -0,0 +1,258 @@
+        private final UpdateAllTask updateAllTask;
+        private final PluginCard card;
+
+        private final boolean isUpdatingAll;
+        private boolean failed;
+
+        UpdateTask(Plugin plugin) {
+            this(plugin, null);
+        }
+
+        UpdateTask(Plugin plugin, UpdateAllTask updateAllTask) {
+            this.plugin = plugin;
+            this.updateAllTask = updateAllTask;
+            this.card = pluginDisplay.getPluginCard(plugin);
+            this.isUpdatingAll = updateAllTask != null;
+            card.setProgressBarMaximum(isUpdatingAll ? 3 : 4);
+        }
+
+        private void waitUntilDone() {
+            try {
+                get();
+            } catch (InterruptedException | ExecutionException e) {
+                e.printStackTrace();
+            }
+        }
+
+        private void publish(PluginCard.UpdateStatus status) {
+            super.publish(status);
+            if (isUpdatingAll) updateAllTask.tickUpdate();
+        }
+
+        @Override
+        protected void process(List<PluginCard.UpdateStatus> chunks) {
+            card.setUpdateProgress(chunks.get(chunks.size() - 1));
+        }
+
+        @Override
+        protected void done() {
+            if (isUpdatingAll) {
+                if (failed) card.setUpdateProgress(PluginCard.UpdateStatus.FAILED);
+                return;
+            }
+
+            if (failed) {
+                card.setUpdateProgress(PluginCard.UpdateStatus.FAILED);
+                // refreshUI() if updating individually and failed to generate ExceptionCards
+                pluginDisplay.refreshUI();
+            }
+            else {
+                plugin.setUpdateStatus(Plugin.UpdateStatus.UP_TO_DATE);
+                card.setUpdateProgress(PluginCard.UpdateStatus.DONE);
+            }
+        }
+
+        @Override
+        protected Void doInBackground() throws Exception {
+            publish(PluginCard.UpdateStatus.STARTING);
+
+            if (plugin.getUpdateDefinition() == null)
+                plugin.setUpdateDefinition(findUpdate(plugin.getDefinition()));
+
+            try (InputStream is = plugin.getUpdateDefinition().download.openConnection().getInputStream()) {
+                publish(PluginCard.UpdateStatus.SAVING_OLD);
+
+                FileUtils.createDirectory(PluginHandler.PLUGIN_OLD_PATH);
+                Files.copy(plugin.getFile().toPath(), PluginHandler.PLUGIN_OLD_PATH.resolve(plugin.getFile().getName()), StandardCopyOption.REPLACE_EXISTING);
+                publish(PluginCard.UpdateStatus.DOWNLOADING);
+
+                FileUtils.createDirectory(PluginHandler.PLUGIN_UPDATE_PATH);
+                Files.copy(is, PluginHandler.PLUGIN_UPDATE_PATH.resolve(plugin.getFile().getName()), StandardCopyOption.REPLACE_EXISTING);
+
+                if (isUpdatingAll) {
+                    publish(PluginCard.UpdateStatus.INDIVIDUALLY_DONE);
+                    plugin.setUpdateStatus(Plugin.UpdateStatus.UP_TO_DATE);
```

Comment:
the up-to-date is set here for all-updates, but on done for single-updates? I think you should make that happen on the same place, probably right before this `if (isUpdatingAll) {`